cpp/ql/test/experimental/query-tests/Security/CWE/CWE-758/semmle/tests/test.cpp

```diff
@@ -0,0 +1,218 @@
+#define NULL (0)
+typedef unsigned long size_t;
+typedef struct {} FILE;
+FILE *ef;
+void *malloc(size_t size);
+size_t strlen(const char *str);
+void free(void *ptr);
+unsigned long copy_from_user (void * to, void * from, unsigned long n);
+int fread(char *buf, int size, int count, FILE *fp);
+int getSize3();
+int getLen3(char *s);
+
+int getSize2()
+{
+  int m;
+  m = getSize3();
+  if (m<0)  return 0;
+  return m;
+}
+
+int getSize1()
+{
+  int m;
+  m = getSize3();
+  if (m<0) m=0;
+  return m;
+}
+int getLen4() {
+  char c[1];
+  fread(c, 1, 1, ef);
+  return (int)*c;
+}
+int getLen2()
+{
+  char str[4] = {0xff, 0xff, 0xff, 0xff};
+  fread(str, 1, 4, ef);
+  return getLen3(str);
+}
+void getLen1(int * len)
+{
+  *len = getLen2();
+}
+
+
+void goodTest1()
+{
+  char *ptr;
+  int len = 10;
+  ptr = (char *)malloc(len); // GOOD
+  if(ptr) {
+	 ptr[len-1] = 0;
+	 free(ptr);
+  }
+}
+void goodTest2(int a,int b)
+{
+  char *ptr; 
+  int len = a+b;
+  ptr = (char *)malloc(len); // GOOD
+  if(ptr) {
+	 ptr[len-1] = 0;
+	 free(ptr);
+  }
+}
+void goodTest3()
+{
+  char *ptr;
+  int len ;
+  len = 10;
+  ptr = (char *)malloc(len); // GOOD
+  if(ptr) {
+	 ptr[len-1] = 0;
+	 free(ptr);
+  }
+}
+void goodTest4(int a,int b)
+{
+  char *ptr;
+  int len;
+  len = a+b;
+  ptr = (char *)malloc(len); // GOOD
+  if(ptr) {
+	 ptr[len-1] = 0;
+	 free(ptr);
+  }
+}
+void goodTest5(int len)
+{
+  char *ptr;
+  if(len == 0)
+    return;
+  ptr = (char *)malloc(len); // GOOD
+  if(ptr) {
+	 ptr[len-1] = 0;
+	 free(ptr);
+  }
+}
+void goodTest6(int len)
+{
+  char *ptr;
+  if(len < 1)
+    return;
+  ptr = (char *)malloc(len); // GOOD
+  if(ptr) {
+	 ptr[len-1] = 0;
+	 free(ptr);
+  }
+}
+void goodTest7(int len)
+{
+  char *ptr = NULL;
+  if (len < 1 || !(ptr = (char *)malloc(len))) // GOOD
+    return;
+	 ptr[len-1] = 0;
+	 free(ptr);
+}
+
+//CVE-2013-6378
+//export and not check
+void badTest1(int len)
+{
+  char *ptr;
+  ptr = (char *)malloc(len); // BAD
+  if(ptr) {
+    ptr[len-1] = 0;
+    free(ptr);
+  }
+}
+void badTest2(int len,char buf[])
```

Comment:
What is the difference between this test case and `badTest1` (other than the existence of a call to `copy_from_user` following the allocation)?

Reply:
differs only in this.
this test case is required, it covers detection when used in a function, size